libhail/docs/aggregators.txt

```diff
@@ -0,0 +1,174 @@
+  }
+
+  class LinearRegressionStateType : public AggStateType {
+    // has no parameters
+    LinearRegressionStateType();
+  };
+
+  class TakeStateType : public AggStateType {
+    TakeStateType(const Type *element_type);
+  };
+
+  class GroupByStateType : public AggStateType {
+    GroupByStateType(const Type *key_type,
+    	             const AggStateType *value_aggstate_type);
+  };
+
+Aggstate types are unrealizable and they do not mention requiredness.
+There is one aggstate type class for each aggregator.
+
+There are two ways of obtaining aggstates: referencing one of the
+aggstates defined at the toplevel (referenced by number in the current
+code), or obtaining an aggstate from a compound aggregator.  Thus, we
+add the following additional IRs for working with aggstate:
+
+  class AggStateRef : public IR {
+    AggStateRef(size_t index);
+  };
+
+  class GroupByStateRef : public IR {
+    GroupByStateRef(IR *aggstate,
+                    IR *key);
+  };
+
+aggstate is the GroupBy's own aggstate.  GroupByStateRef's result is
+the aggstate for key.
+
+RunAgg still manages aggstates and wraps the aggregation process.  It
+look like:
+
+  class RunAgg : public IR {
+    RunAgg(std::vector<AggStateType> signatures,
+	   IR *body,
+	   IR *result);
+  };
+
+There are several IR nodes for interacting with aggstates.  They are
+mostly unchanged, except they now take an `IR *aggstate` instead of an
+integer to refer to the aggstate.
+
+ - InitOp.  InitOp initializes and allocated but uninitialized empty
+   agg state.
+
+  class InitOp : public IR {
+    InitOp(IR *aggstate,
+	   std::vector<IR *> init_args);
+  };
+
+Note, none of these classes have anything that corresponds to
+PhysicalAggSig (which obviously shouldn't belong at the level of the
+IR).
+
+ - InitFromValue.
+
+  class InitFromValue : public IR {
+    InitFromValue(IR *aggstate, *R *aggstate_value);
+  };
+
+Only toplevel aggregators need to be initialized.  Compound
+aggregators are responsible for initializing aggstates that they
+create.  This means that the InitOp for GroupBy(AggX) needs to take
+the InitOp arguments for AggX since it will create instances of AggX's
+aggstate.
+
+ - SeqOp.
+
+  class SeqOp : public IR {
+    SeqOp(IR *aggstate, std::vector<IR *> args);
+  };
+
+ - CombOp.
+
+  class CombOp : public IR {
+    SeqOp(IR *aggstate, IR *src_aggstate);
+  };
+
+ - CombOpValue.
+
+  class ResultOp : public IR {
+    SeqOp(IR *aggstate, IR *aggstate_value);
+  };
+
+ - ResultOp.
+
+  class ResultOp : public IR {
+    SeqOp(IR *aggstate);
+  };
+
+ - AggStateValue.  Converts an aggstate to a Hail value.
+
+  class ResultOp : public IR {
+    SeqOp(IR *aggstate);
+  };
+
+I think of RunAgg as creating (and then cleaning up) a scope of
+aggstates, and managing the aggregation process by running the body
+and result IR.  The signatures are the types of the aggstates bound in
+the scope.
+
+This perspective can be generalized to TableAggregate, allowing us to
```

Comment:
This is a point I want to discuss further. ApplyAggOp has some nice features -- it makes dead code elimination dead simple with respect to aggregators. It's much harder to optimize void-typed IRs, so I want to make sure we have a plan to retain the control we have now.